### DISCUSSION TITLE
On demand prerendering with SvelteKit

### DISCUSSION THREAD

**Discussion by kelvindecosta** [sveltejs/kit#2997] [Ideas]
# Overview

Hello everyone!

SvelteKit offers developers the choice of:

- [Prerendering](https://kit.svelte.dev/docs#appendix-prerendering), which is essentially [Static Site Generation (SSG)](https://kit.svelte.dev/docs#appendix-ssg)
- [Server Side Rendering (SSR)](https://kit.svelte.dev/docs#appendix-ssr) &
- [Client Side Rendering (CSR)](https://kit.svelte.dev/docs#appendix-csr-and-spa)

How can SvelteKit perform prerendering on demand?
In other words, how can a request for a route:

- trigger a build for its page on demand (first ever request)
- serve this freshly built page for all subsequent requests

> **Disclaimer**
>
> I am very new to Svelte and SvelteKit.
> Feel free to correct me where I'm wrong!

## Motivation

Some advantages of this approach:

- Faster deploys, since some pages are not being built at "build time"
- Reduced memory consumption
- Reduced CMS / Third party API querying

These advantages are more pronounced in sites with many static pages.

### Gatsby's DSG

Gatsby v4 came with the ability to use [Deferred Static Generation (DSG)](https://www.gatsbyjs.com/docs/conceptual/rendering-options/#deferred-static-generation-dsg).

A page rendered in DSG mode is built when the first user requests for it.
It is then cached and served to requests made by all users.

### Netlify's DPR

Netlify announced [Distributed Persistent Rendering](https://www.netlify.com/blog/2021/04/14/distributed-persistent-rendering-a-new-jamstack-approach-for-faster-builds/) earlier this year.

With DPR, pages can be [built on demand](https://docs.netlify.com/configure-builds/on-demand-builders/) using appropriate Netlify functions and redirects.

## Conclusion

This approach of building pages only when they are required seems to fit the vision of #transitionalapps.
Although it might not have a huge impact on smaller websites, it can improve both the Dev and User experience of large-scale sites.

Any feedback is greatly appreciated!
Thank you for your time!

**Comment by jwulf0** [comment]
Just thinking - "At first request" seems to be a specific case of "triggered pre-render": Another example of an event where you might want to trigger a fresh prerendering is "When certain data in a CMS changes". So I could see a generic API for prerendering being very useful as well. Or does something like that already exist / is an implementation of my example already possible? (Apart from the obvious possibility to re-build on such CMS changes)

**Comment by stephane-vanraes** [comment]
You might be interested in this <https://github.com/sveltejs/kit/issues/661> 
Essentially, you can add a `max-age` property from your pages which will help with caching.
